clang/lib/3C/3C.cpp

```diff
@@ -741,3 +750,89 @@ void _3CInterface::invalidateAllConstraintsWithReason(
     delete (ToDelCons);
   }
 }
+
+bool _3CInterface::makeSinglePtrNonWild(ConstraintKey TargetPtr) {
+  std::lock_guard<std::mutex> Lock(InterfaceMutex);
+  CVars RemovePtrs;
+  RemovePtrs.clear();
+
+  auto &PtrDisjointSet = GlobalProgramInfo->getInterimConstraintState();
+  auto &CS = GlobalProgramInfo->getConstraints();
+
+  CVars OldWildPtrs = PtrDisjointSet.AllWildAtoms;
+
+  //Delete the constraint that makes the target non-wild
+  VarAtom *VA = CS.getOrCreateVar(TargetPtr, "q", VarAtom::V_Other);
+  Geq NewE(VA, CS.getWild(), ReasonLoc());
+  Constraint *OriginalConstraint = *CS.getConstraints().find(&NewE);
+  CS.removeConstraint(OriginalConstraint);
+  VA->getAllConstraints().erase(OriginalConstraint);
+
+  delete(OriginalConstraint);
+
+  // Reset the constraint system.
+  CS.resetEnvironment();
+
+  // Solve the constraints.
+  //assert (CS == GlobalProgramInfo->getConstraints());
+  runSolver(*GlobalProgramInfo, FilePaths);
+
+  // Compute new disjoint set.
+  assert(CStateisclear == true);
+  GlobalProgramInfo->computeInterimConstraintState(FilePaths);
+
+  // Get new WILD pointers.
+  CVars &NewWildPtrs = PtrDisjointSet.AllWildAtoms;
+
+  // Get the number of pointers that have now converted to non-WILD.
+  std::set_difference(OldWildPtrs.begin(), OldWildPtrs.end(),
+                      NewWildPtrs.begin(), NewWildPtrs.end(),
+                      std::inserter(RemovePtrs, RemovePtrs.begin()));
+
+  return !RemovePtrs.empty();
+}
+
+bool _3CInterface::invalidateWildReasonGlobally(ConstraintKey PtrKey) {
+  std::lock_guard<std::mutex> Lock(InterfaceMutex);
+
+  CVars RemovePtrs;
+  RemovePtrs.clear();
+
+  auto &PtrDisjointSet = GlobalProgramInfo->getInterimConstraintState();
+  auto &CS = GlobalProgramInfo->getConstraints();
+
+  CVars OldWildPtrs = PtrDisjointSet.AllWildAtoms;
+
+  // Delete ALL the constraints that have the same given reason.
+  VarAtom *VA = CS.getOrCreateVar(PtrKey, "q", VarAtom::V_Other);
+  Geq NewE(VA, CS.getWild(), ReasonLoc());
```

Comment:
Geq does mean ">=". And nothing is greater than wild. Since we already have Geq that is used in other constraints, using the same here makes it simpler (Instead of creating a new type just to equate with wild).

Reply:
Could you add a comment to that effect?